### DISCUSSION TITLE
Welcome to geppy Discussions!

### DISCUSSION THREAD

**Discussion by ShuhuaGao** [ShuhuaGao/geppy#39] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by OneWingAngel** [comment]
Greetings, my dear friends. I'm trying to learn this geppy recently, but I'm suffering from a problem. 
In the part of funtions, I'm trying to use both built-in functions, such as + - * /, as well as customized functions, such as mean(), correlation(), ect.
However, my problem is, in some functions, I only need float numbers inputs, such as + -, but in other functions, I need to use pandas Series inputs, such as mean() and skew(). I'm not sure how to distinguish different types of inputs, and how to write this kind of codes.
[GEPPY Problem.txt](https://github.com/ShuhuaGao/geppy/files/9848165/GEPPY.Problem.txt)
For example, in this code, in the function of correlation, I want a1 and b1 to be a pandas Series, and days to be a float, but I found their types is random, each of them will be either a Series or a float, which I cannot control. Is there any way to handle this problem? Thank you guys so much.

**Comment by ShuhuaGao** [comment]
Sorry. `geppy` is not strongly typed GP. Currently, it cannot meet your requirement. You may check `deap` though.

**Reply by OneWingAngel** [reply]
Thank you my friend, I will try deap.

**Comment by renzhitian** [comment]
Hi Shuhua,

Thank you very much for your geppy package. It helped me a lot when I learnt GEP. 

Currently, I extend it with a local optimizer to deal with ephemeral constants, and received very good results for algebric expressions.

However, I also find some aspects concerning to potentially improvement of GEP. I list them as follows.
1, When it comes to real physics, if nondimensionization is inappropriate, expressions are usually going with dimensions (say mass, length, time). How Can GEP be dimension-aware?
2, For real dataset and features (including several extra features that are not significantly related to dataset), generally, the final best evolutionary result are too long with the risk of overfitting. How can we reduce the expression length as the way sparse or elastic regression usually does?

How do you think we can resolve these two problems of GEP?

Best wishes,
Renzhi

**Comment by ShuhuaGao** [comment]
Hi, @renzhitian , thanks for your feedback.

1. I have no clear idea. In fact, I believe Julia is a better language for GP, especially strongly-typed and unitful ones (like the one you mentioned). See https://painterqubits.github.io/Unitful.jl/stable/. Nonetheless, currently, I have no plan to rewrite geppy in Julia. Contributions are welcome.
2. This is a common issue called _bloating_ in the literature. You need to add some normalization terms, e.g., penalize the complexity of the model. For example, you may set `fitness = MSE + w * (number of nodes in the tree)`, where `w` is a tunable weight factor. There are many other techniques. Please search GP bloating.

**Reply by renzhitian** [reply]
Hi @ShuhuaGao,
Thanks for your comments. I noted Julia down for later searching and exploring since I have no basics yet. If I have any progress on GEP with Julia shortly, I will let you know~
Regarding GP bloating, I viewed some literature and am thinking maybe regularization used in general symbolic regression is also worth a try in GEP.

**Reply by ShuhuaGao** [reply]
https://github.com/MilesCranmer/SymbolicRegression.jl may be interesting, though it is not for GEP.

**Reply by renzhitian** [reply]
Thank you very much!

**Reply by renzhitian** [reply]
Hi @ShuhuaGao,
Sorry for this late reply. finally, I did not turn to julia, because my supervisor and I, since we are not programming experts, both believe a physics-related problem in one language cannot be solved by changing a language without extra effort. Finally, I used a Python package called "pint" to filter the genes with invalid units.

**Reply by maxreiss123** [reply]
Hi, thank you for your package, and I apologize for the late reply. I faced a similar problem according to the dimensionality, so I tried to tackle it in Python (with geppy). Unfortunately, every correction made during the exploration of each expression comes with extra time, significantly expanding the runtime of the Python implementation. Therefore, I decided to re-implement the Gene Expression Programming in Julia. https://github.com/maxreiss123/GeneExpressionProgramming.jl

**Comment by moosafly** [comment]
Hello, my friend. When I use the second case of geppy, I find that the code that generates the treemap reports an error, why is that?
